theseus/optimizer/nonlinear/nonlinear_least_squares.py

```diff
@@ -4,18 +4,58 @@
 # LICENSE file in the root directory of this source tree.
 
 import abc
-from typing import Any, Dict, Optional, Type
+import warnings
+from typing import Any, Callable, Dict, NoReturn, Optional, Tuple, Type, Union
 
 import torch
 
 from theseus.core import Objective
-from theseus.optimizer import Linearization
-from theseus.optimizer.linear import LinearSolver
-from theseus.optimizer.linear.dense_solver import CholeskyDenseSolver
+from theseus.optimizer import Linearization, OptimizerInfo
+from theseus.optimizer.linear import (
+    LinearSolver,
+    CholeskyDenseSolver,
+    LUCudaSparseSolver,
+)
+from .nonlinear_optimizer import (
+    BackwardMode,
+    NonlinearOptimizer,
+    NonlinearOptimizerInfo,
+    NonlinearOptimizerStatus,
+)
 
-from .nonlinear_optimizer import NonlinearOptimizer
 
+EndIterCallbackType = Callable[
+    ["NonlinearOptimizer", NonlinearOptimizerInfo, torch.Tensor, int], NoReturn
+]
 
+
+# Base class for all optimizers for NLLS problems,
+# providing the skeleton of the
+# optimization loop. Subclasses need to implement the following method:
+#
+#   - `compute_delta`: returns a descent direction given the current values
+#     of the objective's optimization vars.
+#
+# Optionally, they can also provide the following methods:
+#
+#   - `reset`: resets any internal state needed by the optimizer.
+#   - `_complete_step`: called at the end of an optimization step, but before
+#     optimization variables are updated. Returns batch indices that should not
+#     be updated (e.g., if the step is to be rejected).
+#
+# The high level logic of a call to optimize is as follows:
+#
+# prev_err = objective.error_squared_norm() / 2
+# do optimization loop:
+#    1. compute delta
+#    2. step(delta, prev_err)
+#           2.1. Store current optim var tensors in tmp_optim_vars containers
+#           2.2. Retract all tmp_optim_vars given delta
+#           2.3. Evaluate new error
+#           2.4. reject_indices = self._complete_step(delta, new_err, prev_err)
+#           2.5. Update objective's optim var containers with retracted values,
+#                ignoring indices given by `reject_indices`
+#    3. Check convergence
 class NonlinearLeastSquares(NonlinearOptimizer, abc.ABC):
```

Comment:
Shouldn't this class just define `_error_metric()`? rest such as `_optimize_impl()` should still be left abstracted for a new subclass?

Reply:
Maybe? I guess depends on how we resolve [this](https://github.com/facebookresearch/theseus/pull/440#discussion_r1088356231).

Reply:
Sounds good.